Monika-After-Story/MonikaModDev
# Black Bow + defaulting mux types based on acs type

# Key changes
Adds Black Bow from multi.

Also adds a mapping zone so we can specify acs types to have a default list of mux types. This means we don't have to update every single ribbon when we want to add mux types for ribbons. If an ACS needs extra mux types, that should still be done the normal way.

The map is located at the end of `zz_spriteobjects`. In `ch30_reset`, all ACS sprites are checked, and if their `acs_type` is in the map, the corresponding `mux_types` will be added to the ACS. **this does NOT overwrite ACS mux types set manually in the ACS object or sprite json**

All ribbon selector dialogue has been changed to "Hair tie". There is also an option to change the name of a remover object. For ribbons, its "Basic Hair Band".

## new selector mechanics
* validation and unlocking of selector topics to the `PROMPT_MAP` instead of the `GRP_TOPIC_MAP`.
* `PROMPT_MAP` now accepts a `_rule` item that can take a function. This function should return True if the selector can be unlocked, False if not.
* `PROMPT_MAP` now accepts `_min-items`, which functions the same as the 2nd arg of `GRP_TOPIC_MAP` entries.
* `mas_selspr.check_prompt` - given a prompt key (aka group), check if the prompt's `_rule` passes. This returns True for prompts without a rule.
* `mas_selspr.get_minitems` - given a prompt key, get the `_min-items` property. this also accepts a `defval`.
* `mas_selspr.lock_prompt` - locks selector associated with the given prompt key
* `mas_selspr.unlock_prompt` - unlocks selector associated with the given prompt key
* new param `allow_lock` added to `mas_filterUnlockGroup`. If True, then the function can also lock a selector topic if it does not pass rules.

## New exprops:
* `ribbon-like` - add key to ACS that should be considered a ribbon-like ACS. this means they are removed and restored like a ribbon when switching to baked outfits. Also ribbon-like stuff cannot be used with non-ribbon hairstyles.

## New sprite json props:
* `dlg_desc` - user friendly name of the accessory to be used in dialogue. should be all lowercase
* `dlg_plural` - boolean that says whether or not the accessory should be used with plural descriptors.
* these 2 properties are stored as `dlg_desc` and `dlg_plur` in the `MASAccessory` object.

# Testing
* verify black bow looks ok. 
* verify the bow is removed when using baked outfits and non-ribbon hairs.
* verify all acs selectors are unlocked/locked correctly.
* verify generic gift reaction for black bow says "this black bow. I really appreciate it!"
* verify that switching between lanval lime ribbons and the blackbow works (aka both ACS are not worn at the same time). The lanvallime ribbon JSONs were not changed, so they still reference a `mux_type` of only `ribbon`. They are getting the `bow` mux type via the default mux type system described above. (Also only some of the built-in ribbons have the bow in mux type, I got tired halfway through changing them and made the default mux type system)

you can run `dev_mas_unlock_all_sprites` to unlock all selectables.
also `dev_mas_clear_spritegift` to clear all data related to a sprite gift via giftname, but it may not work 100% unless you also launch the game without the json to fully clear all data.
